core/trino-main/src/test/java/io/trino/execution/scheduler/TestScaledWriterScheduler.java

```diff
@@ -0,0 +1,405 @@
+import io.trino.sql.planner.PartitioningScheme;
+import io.trino.sql.planner.PlanFragment;
+import io.trino.sql.planner.Symbol;
+import io.trino.sql.planner.plan.PlanFragmentId;
+import io.trino.sql.planner.plan.PlanNodeId;
+import io.trino.sql.planner.plan.TableScanNode;
+import io.trino.testing.TestingMetadata;
+import io.trino.util.FinalizerService;
+import org.joda.time.DateTime;
+import org.testng.annotations.Test;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Optional;
+import java.util.OptionalLong;
+
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static io.trino.execution.TestingRemoteTaskFactory.TestingRemoteTask;
+import static io.trino.spi.type.VarcharType.VARCHAR;
+import static io.trino.sql.planner.SystemPartitioningHandle.SINGLE_DISTRIBUTION;
+import static io.trino.sql.planner.SystemPartitioningHandle.SOURCE_DISTRIBUTION;
+import static io.trino.testing.TestingHandles.TEST_TABLE_HANDLE;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.testng.Assert.assertEquals;
+
+public class TestScaledWriterScheduler
+{
+    private static final PlanNodeId TABLE_SCAN_NODE_ID = new PlanNodeId("plan_id");
+    private static final InternalNode NODE_1 = new InternalNode("node-1", URI.create("http://127.0.0.1:11"), new NodeVersion("version-1"), false);
+    private static final InternalNode NODE_2 = new InternalNode("node-2", URI.create("http://127.0.0.1:12"), new NodeVersion("version-1"), false);
+
+    @Test
+    public void testGetNewTaskCountWithUnderutilizedNonSkewedTasks()
+    {
+        TaskInfo taskInfo1 = buildTaskInfo(false, 12345L);
+        TaskInfo taskInfo2 = buildTaskInfo(false, 12345L);
+        TaskInfo taskInfo3 = buildTaskInfo(true, 12345L);
+
+        ScaledWriterScheduler scaledWriterScheduler = new ScaledWriterScheduler(
+                new TestingStageExecution(createFragment()),
+                () -> ImmutableList.of(taskInfo1, taskInfo2, taskInfo3),
+                () -> ImmutableList.of(taskInfo1.getTaskStatus(), taskInfo2.getTaskStatus(), taskInfo3.getTaskStatus()),
+                new UniformNodeSelectorFactory(
+                        new InMemoryNodeManager(NODE_1, NODE_2),
+                        new NodeSchedulerConfig().setIncludeCoordinator(true),
+                        new NodeTaskMap(new FinalizerService())).createNodeSelector(testSessionBuilder().build(), Optional.empty()),
+                newScheduledThreadPool(10, threadsNamed("task-notification-%s")),
+                DataSize.of(32, DataSize.Unit.MEGABYTE),
+                1);
+
+        scaledWriterScheduler.schedule();
+
+        assertEquals(scaledWriterScheduler.schedule().getNewTasks().size(), 0);
+    }
+
+    @Test
+    public void testGetNewTaskCountWithUOverutilizedNonSkewedTasks()
```

Comment:
`testScheduleWithOverutilizedNonSkewedTasks`

Reply:
I think we can remove this test since the last test already cover this case.

Reply:
I think it's still meaningful for the case when all tasks are evenly distributed(no skewed tasks). Will fix the typo.